python/GafferUI/Bookmarks.py

```diff
@@ -300,8 +300,10 @@ def __save( self ) :
 		f.write( "# Do not edit this file - it will be overwritten.\n\n" )
 
 		f.write( "import Gaffer\n" )
-		f.write( "import GafferUI\n" )
-
+		f.write( "\n" )
+		f.write( "if 'gui' in application.parameters() and application['gui'].getTypedValue() :\n" )
```

Comment:
This breaks bookmark loading for the main app and `gaffer view` etc, because they don't have a `gui` parameter. I'm also reluctant to query `application.parameters()` from a file we'll be saving everywhere and have no control over, because my intention is to replace the parameters with plugs at some point.

If we were doing this again, I think I'd have had different apps for `gaffer op -gui 1` and `gaffer op -gui 0`, but I don't suppose that it an option for you now? It would certainly be fine with me, because we don't ship GafferCortex anyway.

Perhaps the simplest solution is to check `self.__applicationRoot().name() == "op"` here, and only write out the conditional stuff if that matches. Then we'll keep the files clean for all other apps, and only have a backwards compatibility issue for a single app when we switch to plugs in future.

Reply:
Oh shoot, I totally failed to test the main app... note the dispatch app also needs this treatment as it also has the “gui” arg

Reply:
I think a single app that has a gui/no-gui mode makes sense, `dispatch` being the current "official" example and `op` being the legacy example.

I could

- Add an explicit `self.__applicationRoot().typeName() in ( "op", "dispatch" )` check that would need updating anytime we add another app (and would be quite hard to update for 3rdParty extension apps).
- Add `Application.guiMode()` (or similar) and implement it to return `true` or `false` as appropriate for each app (default implementation `true` I presume).

What do you prefer?

Reply:
> I think a single app that has a gui/no-gui mode makes sense

Can you explain why? The best reason I can come up with is the shared arguments and setup, but that could be achieved without cramming everything into one app.

> What do you prefer?

For now I prefer `typeName() in ( "op", "dispatch" )` - it's a hack, but it'll do the job and I don't want to commit to a more official mechanism in a rush.

Reply:
> Can you explain why? The best reason I can come up with is the shared arguments and setup, but that could be achieved without cramming everything into one app.

I think shared arguments and setup are a good reason. As a code-client of an app, instantiating a `dispatch` app, setting arguments, handing that app through various configs and other pipeline code (which may or may not ask the user for GUI input), and then running it feels more natural than instantiating something else (`dispatch-arguments`?) and passing that around (and I don't even know how one would optionally display it to a user this way).

Anyhow, the hack approach is fine for us for now, and the more long term approach certainly isn't a priority, so I don't think we need an answer to that question right now.

Reply:
Addressed in 7954bf5f15466b615bbf2650a426e0ec93cb787b

Reply:
> I think shared arguments and setup are a good reason. 

They're a good goal, but a single app that does everything is not the only way of achieving it.

> As a code-client of an app, instantiating a dispatch app, setting arguments, handing that app through various configs and other pipeline code (which may or may not ask the user for GUI input), and then running it feels more natural than instantiating something else (dispatch-arguments?) and passing that around (and I don't even know how one would optionally display it to a user this way).

All of those could be achieved by a base class app which defines the plugs, and then a derived app for command line and gui usage. Or an app for command line usage and a gui one derived from it.